Add renameParam and renameCoupledVar APIs

lindsayad opened the pull request:
Closes #11117 #23259

Checked renaming for:
 - [x] basic parameter (e.g. Real or something)
 - [x] coupled field var
 - [x] coupled scalar var
 - [x] functor
 - [x] material property
 - [x] postprocessor

Any interface that queries for input parameter names using `params.get<T>("code_param_name")` are automatically supported. Coupled variables were a little bit tougher since in the constructors we query for the coupled variable names held by the input parameters object, and these do not include the old, renamed variable names. So if in base class user code, the user asks for the variable corresponding to the old variable name, then we have to manually map to the new name before proceeding.

lindsayad committed:
Add renameParam and renameCoupledVar APIs

Closes #23259

lindsayad committed:
Test

YaqiWang commented on framework/src/utils/InputParameters.C:
```diff
@@ -1195,3 +1230,61 @@ InputParameters::varName(const std::string & var_param_name,
+
+  _last_checked_rename.first = name_in;
+  if (auto it = _old_to_new_name.find(name_in); it != _old_to_new_name.end())
+    _last_checked_rename.second = it->second;
```
Do we issue a deprecation message here?

dschwen replied:
Is this for deprecation or for derived classes to provide more meaningful parameter names?

lindsayad replied:
The latter

roystgnr replied:
Would it be useful for deprecation, if we added a flag to distinguish the two use cases and avoid false positive deprecation messages?  Deprecating poor initial choices of parameter names was the first use case that came to my mind too.

lindsayad replied:
I can think about how to integrate this into our existing deprecation APIs

YaqiWang commented on framework/src/utils/InputParameters.C:
```diff
@@ -1195,3 +1230,61 @@ InputParameters::varName(const std::string & var_param_name,
+InputParameters::renameParam(const std::string & old_name,
+                             const std::string & new_name,
+                             const std::string & new_docstring)
+{
```
I guess if we want to remove the old name completely for a parameter in the future, we can just edit the name in `addParam` functions and the docstring and remove this function?

YaqiWang reviewed (changes requested):
Looks good. This possibly can be used to close #11117. It is not necessary to have my suggested interface because we have several variants of functions for adding parameters.

lindsayad commented:
My aim here is to support application developers overriding the general parameter names of base classes. Base classes still use the general name in their code

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1319897/) on 0a761b9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23260/site)

This comment will be updated on new commits.

lindsayad committed:
Prevent thread races

lindsayad committed:
Add a couple new error checks

Add a template parameter to `renameParam`. This enables
- Checking consistent template types with the old name
- Making sure that the user is requesting to rename a parameter
  for a supported system

lindsayad committed:
Wrap Parameters::get

lindsayad committed:
Wrap Parameters::have_parameter

lindsayad committed:
No longer need conversions in most interface classes

lindsayad committed:
Revert "Add a couple new error checks"

This reverts commit 6747fa13d6a8437b7c43fa033c3cf51157102fb4.

With the generic `get` and `have_parameter` wrappers we should
have most interfaces automatically supported for parameter
renaming. Hence we can remove the `std::is_same` checks

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1320294/) on f33c3fa : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1320303/) on f33c3fa wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>26e8ed</th>
      <th colspan=4>#23260 f33c3f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.62%</td>
      <td>84.65%</td>
      <td>+0.03%</td>
      <td>94.66%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81975</td>
      <td>82078</td>
      <td>+103</td>
      <td>124</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14901</td>
      <td>14887</td>
      <td>-14</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23260/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23260/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>26e8ed</th>
      <th colspan=4>#23260 f33c3f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.44%</td>
      <td>89.46%</td>
      <td>+0.02%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26461</td>
      <td>26467</td>
      <td>+6</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3123</td>
      <td>3117</td>
      <td>-6</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23260/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23260/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23260/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud reviewed (commented):
is there any framework object this is used on? I d like to check how it looks in the docs.

I can also build locally if not

lindsayad commented:
> is there any framework object this is used on?

No, not currently although I believe there is in #22376 which is the reason I took this task on. In general I would imagine that there won't be much renaming in the framework; more so in the modules/applications.

lindsayad committed:
Exploring using rename machinery for deprecation

lindsayad committed:
Remove use of addDeprecatedCoupledVar in NS

lindsayad committed:
Account for renamed parameters in applyParameter

lindsayad committed:
Remove use of addDeprecatedCoupledVar in GrainGrowthAction

This curiously was causing a deprecated failure when running a
NS input using an NS action with the combined executable.

Some thermal hydraulics failures also included GrainGrowthAction
in the deprecation stack trace

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1342820/) on 2dbebe9 : invalidated by @lindsayad

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1342796/) on 2dbebe9 : invalidated by @lindsayad

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1342796/) on 2dbebe9 : invalidated by @lindsayad

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1342819/) on 2dbebe9 : invalidated by @lindsayad

joshuahansel reviewed (changes requested):
This is news-worthy - add newsletter entry.

joshuahansel commented:
I've found myself wanting to use this in my code several times in the past week 🙌

lindsayad commented:
Love that!

lindsayad committed:
Increase max_time for appsyntax

GiudGiud commented on framework/include/materials/CompositeTensorBase.h:
```diff
@@ -106,12 +106,8 @@ CompositeTensorBase<T, U>::validParams()
-  // TODO Make required once deprecation is handled, see #19119
-  params.addCoupledVar("coupled_variables", "variable dependencies for the prefactor");
-
+  params.addRequiredCoupledVar("args", "variable dependencies for the prefactor");
```
eh doesnt matter

lindsayad replied:
The deprecation work piggy-backed off the rename work in which it is a prerequisite that the old parameter exist when the rename is attempted. Because we perform a map lookup for the old parameter, move its metadata over to the new parameter, and erase the old parameter.

For a developer performing deprecation, they will have already had this line in their code when they come to perform the deprecation. Then the deprecation is only a single new statement; they do not have to modify any statements. Of course when you finally remove the deprecated param you have to modify one statement, which modification could be avoided with your suggestion, and remove the other.

So the total developer work is the same ... but in terms of back-end implementation it is easier if I can reuse the same machinery for both renaming and deprecation

GiudGiud commented on framework/src/kernels/ParsedODEKernel.C:
```diff
@@ -25,17 +25,11 @@ ParsedODEKernel::validParams()
+  params.addRequiredCustomTypeParam<std::string>(
+      "function", "FunctionExpression", "function expression");
+  params.deprecateParam("function", "expression", "02/07/2024");
+  params.addCoupledVar("args", "additional coupled variables");
```
```suggestion
  params.addCoupledVar("args", "Scalar variables coupled in the parsed expression.");
```

GiudGiud commented on framework/src/utils/InputParameters.C:
```diff
@@ -1122,8 +1166,9 @@ InputParameters::checkParamName(const std::string & name) const
+  const auto name = checkForRename(name_in);
   auto it = _params.find(name);
   if (it != _params.end())
     return it->second._ignore;
```
there's two get routines below that could use a checkforrename ?

lindsayad replied:
They only use internal data. For the incoming `group` argument, I'm not doing anything about renaming parameter groups here.

lindsayad committed:
Add newsletter entry

GiudGiud reviewed (changes requested):
looks real good. 
I was actually going to add a check in parsed objects for either "f_name" or "property_name" being passed (formerly required) because someone just thought they could wing it with the name of the Material instead of the property_name. But this is better

GiudGiud commented on modules/doc/content/newsletter/2023/2023_02.md:
```diff
@@ -35,6 +35,69 @@ This will apply the coordinate transformation (rotations and scaling defined in
+
+The machinery underlying the new `rename` and `deprecate` APIs allows object
+code to simply use the new names as opposed to checking for both deprecated/base-class and
+new/blessed parameters. For instance this code in the `ParsedAux` constructor
```
making fun of my code :P

lindsayad replied:
haha no shade intended!

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1033,8 +1106,12 @@ InputParameters::setHelper(const std::string & /*name*/)
-InputParameters::set(const std::string & name, bool quiet_mode)
+InputParameters::set(const std::string & name_in, bool quiet_mode)
 {
+  const bool is_function = name_in == "function";
+  const auto name = checkForRename(name_in);
+  libmesh_ignore(is_function);
```
oh whoops. Nice catch. This was a debugging line

lindsayad committed:
Apply suggestions from code review

- typo fix
- more detailed documentation string
- More doxygen
- Remove debugging variable

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1358280/) on bfc2237 : invalidated by @lindsayad

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1016,6 +1086,9 @@ class InputParameters : public Parameters
   /// A map from deprecated coupled variable names to the new blessed name
   std::unordered_map<std::string, std::string> _new_to_deprecated_coupled_vars;
 
+  std::map<std::string, std::pair<std::string, std::string>> _old_to_new_name_and_dep;
```
done

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -858,6 +859,63 @@ class InputParameters : public Parameters
+  template <typename T>
+  bool have_parameter(std::string_view name) const;
+
+  std::vector<std::string> paramAliases(const std::string & param_name) const;
```
done

lindsayad commented on framework/include/kernels/MaterialDerivativeTestKernelBase.h:
```diff
@@ -67,11 +67,8 @@ MaterialDerivativeTestKernelBase<T>::validParams()
-                                 "args is deprecated, use 'coupled_variables' instead");
-  // TODO Make required once deprecation is handled, see #19119
-  params.addCoupledVar("coupled_variables", "List of variables the material property depends on");
+  params.addRequiredCoupledVar("args", "List of variables the material property depends on");
```
see https://github.com/idaholab/moose/pull/23260#discussion_r1110332880

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1358280/) on bfc2237 : invalidated by @lindsayad
